langserver/loader.go

```diff
@@ -306,3 +312,47 @@ func isMultiplePackageError(err error) bool {
 	_, ok := err.(*build.MultiplePackageError)
 	return ok
 }
+
+var freeOSMemoryOnce sync.Once
+
+// freeOSMemory spawns a goroutine (only once) which invokes
+// runtime/debug.FreeOSMemory() more aggressively than the runtime default of
+// 5 minutes after GC.
+//
+// There is a long-standing known issue with Go in which memory is not returned
+// to the OS aggressively enough[1], which coincidently harms our application
+// quite a lot because we perform so many short-burst heap allocations during
+// the type-checking phase.
+//
+// This function should only be invoked in editor mode, not in sourcegraph.com
+// mode, because users running the language server as part of their editor
+// generally expect much lower memory usage. In contrast, on sourcegraph.com we
+// can give our servers plenty of RAM and allow Go to consume as much as it
+// wants. Go does reuse the memory not free'd to the OS, and as such enabling
+// this does _technically_ make our application perform less optimally -- but
+// in practice this has no observable affect in editor mode.
+//
+// The end effect of performing this is that repeating "hover over code" -> "make an edit"
+// 10 times inside a large package like github.com/docker/docker/cmd/dockerd:
+//
+//
+// 	| Real Before | Real After | Real Change | Go Before | Go After | Go Change |
+// 	|-------------|------------|-------------|-----------|----------|-----------|
+// 	| 7.61GB      | 4.12GB     | -45.86%     | 3.92GB    | 3.33GB   | -15.05%   |
+//
+// Where `Real` means real memory reported by OS X Activity Monitor, and `Go`
+// means memory reported by Go as being in use.
+//
+// TL;DR: 46% less memory consumption for users running with the vscode-go extension.
+//
+// [1] https://github.com/golang/go/issues/14735#issuecomment-194470114
+func freeOSMemory() {
```

Comment:
Yep, that's true. The reason I did it this way is to keep it close to the problem (type-checking) and because I plan to use the `editor` bool in this same location in another change soon.